File: 2022/03/02/guessNumber.ts
/**
 * https://leetcode.com/problems/guess-number-higher-or-lower/
 * @author Andrew Jarombek
 * @since 3/2/2022
 */

const num = 6;

function guessFactory(num: number): (n: number) => number {
    return (n: number): number => {
        if (n == num) {
            return 0;
        } else if (n < num) {
            return -1;
        } else {
            return 1;
        }
    }
}

const guess = guessFactory(num);

function guessNumber(n: number): number {
    let low = 1;
    let high = n;

    while (low <= high) {
        const mid = Math.floor((low + high) / 2);
        const result = guess(mid);

        if (result == 0) {
            return mid;
        } else if (result == 1) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }

    return high;
}

File: 2022/07/18/java/GuessNumber.java
/**
 * https://leetcode.com/problems/guess-number-higher-or-lower/
 * @author Andrew Jarombek
 * @since 7/18/2022
 */

public class GuessNumber {
    private static int guess(int num) {
        if (num == 6) {
            return 0;
        } else if (num < 6) {
            return 1;
        } else {
            return -1;
        }
    }

    public static int guessNumber(int n) {
        long low = 1;
        long high = n;

        while (low <= high) {
            long mid = (low + high) / 2;
            int result = guess((int) mid);

            if (result == 0) {
                return (int) mid;
            } else if (result == -1) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return (int) low;
    }

    public static void main(String[] args) {
        assert guessNumber(10) == 6;
    }
}
